File: frontend/src/types/alert.ts
export interface MitreTechnique {
  technique_id: string
  technique_name: string
  tactic: string
}

export interface VTResult {
  indicator: string
  malicious_count: number
  suspicious_count?: number
  total_engines?: number
  detection_ratio?: number
  threat_categories?: string[]
  permalink?: string
  cached?: boolean
  error?: string
}

export interface AbuseResult {
  ip: string
  abuse_score: number
  isp?: string
  usage_type?: string
  country_code?: string
  total_reports?: number
  categories?: string[]
  permalink?: string
  cached?: boolean
}

export interface FeedMatch {
  indicator: string
  feed_name: string
  confidence?: number
  source?: string
}

export interface IocGroup {
  ips: string[]
  domains: string[]
  hashes: string[]
  cves: string[]
}

export interface Enrichment {
  /** Structured IOC groups written by the backend enrichment pipeline. */
  iocs: IocGroup
  vt_results: VTResult[]
  abuse_results: AbuseResult[]
  feed_matches: FeedMatch[]
  risk_modifier: number
  summary: string
}

export type AlertStatus =
  | 'open'
  | 'in_progress'
  | 'resolved'
  | 'false_positive'

export interface Alert {
  id: string
  title: string
  source: string
  severity: 'Low' | 'Medium' | 'High' | 'Critical'
  category: string
  status: AlertStatus
  assigned_to: string | null
  asset_id: string | null
  user_id: string | null
  source_ip: string | null
  dest_ip: string | null
  mitre_techniques: MitreTechnique[]
  risk_score: number | null
  risk_level: string | null
  confidence: number | null
  incident_group_id: string | null
  ai_summary: string | null
  ai_key_facts: Record<string, unknown> | null
  ai_remediation: Record<string, unknown> | null
  ai_next_steps: Record<string, unknown> | null
  enrichment: Enrichment | null
  created_at: string
  updated_at: string
  raw?: Record<string, unknown>
  decision?: {
    risk_score?: number
    risk_level?: string
    confidence?: number
    explanation?: { items?: string[] }
    recommended_actions?: { items?: string[] }
    mitre_techniques?: unknown
    correlation_rule_triggered?: string | null
  }
}

export interface AlertsResponse {
  alerts: Alert[]
  total: number
  page: number
  limit: number
}

export interface AlertFiltersParams {
  severity?: string
  status?: string
  source?: string
  category?: string
  search?: string
  assigned_to?: string
  date_from?: string
  date_to?: string
  page?: number
  limit?: number
}

File: app/models/alert.py
"""Alert model for security events."""

import enum
from datetime import datetime

from sqlalchemy import DateTime, Enum, Float, Integer, String, Text, func
from sqlalchemy.types import JSON
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class AlertSeverity(str, enum.Enum):
    LOW = "Low"
    MEDIUM = "Medium"
    HIGH = "High"
    CRITICAL = "Critical"


class AlertStatus(str, enum.Enum):
    OPEN = "open"
    IN_PROGRESS = "in_progress"
    RESOLVED = "resolved"
    FALSE_POSITIVE = "false_positive"


class Alert(Base):
    __tablename__ = "alerts"

    id: Mapped[str] = mapped_column(String(36), primary_key=True)
    title: Mapped[str] = mapped_column(String(512), nullable=False)
    source: Mapped[str] = mapped_column(String(100), nullable=False, index=True)
    severity: Mapped[AlertSeverity] = mapped_column(
        Enum(AlertSeverity), nullable=False, index=True
    )
    category: Mapped[str] = mapped_column(String(64), nullable=False)
    status: Mapped[AlertStatus] = mapped_column(
        Enum(AlertStatus), nullable=False, default=AlertStatus.OPEN, index=True
    )
    assigned_to: Mapped[str | None] = mapped_column(String(256), nullable=True)
    asset_id: Mapped[str | None] = mapped_column(String(256), nullable=True)
    user_id: Mapped[str | None] = mapped_column(String(256), nullable=True)
    source_ip: Mapped[str | None] = mapped_column(String(45), nullable=True)
    dest_ip: Mapped[str | None] = mapped_column(String(45), nullable=True)
    mitre_techniques: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    raw: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    ai_summary: Mapped[str | None] = mapped_column(Text, nullable=True)
    ai_key_facts: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    ai_affected: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    ai_evidence: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    ai_remediation: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    ai_next_steps: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    risk_score: Mapped[int | None] = mapped_column(Integer, nullable=True)
    risk_level: Mapped[str | None] = mapped_column(String(32), nullable=True)
    confidence: Mapped[float | None] = mapped_column(Float, nullable=True)
    incident_group_id: Mapped[str | None] = mapped_column(
        String(36), nullable=True, index=True
    )
    correlated_alert_ids: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    enrichment: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), index=True
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )
